lib/util.ts

```diff
@@ -82,38 +108,71 @@ export class DetailContext implements IContext {
     if (best && best._score > 0) {
       this._propNames.push(...best._propNames);
       this._messages.push(...best._messages);
+      this._forks.push(...best._forks);
     }
   }
 
   public getError(path: string): VError {
-    const msgParts: string[] = [];
-    for (let i = this._propNames.length - 1; i >= 0; i--) {
-      const p = this._propNames[i];
-      path += (typeof p === "number") ? `[${p}]` : (p ? `.${p}` : "");
-      const m = this._messages[i];
-      if (m) {
-        msgParts.push(`${path} ${m}`);
-      }
-    }
-    return new VError(path, msgParts.join("; "));
+    const fullMessage = this.getErrorDetails(path)
+        .map(d => `${d.path} ${d.message}`)
+        .join("; ");
+    return new VError(path, fullMessage);
   }
 
-  public getErrorDetail(path: string): IErrorDetail|null {
+  public getErrorDetail(path: string): IErrorDetail {
+    return this.getErrorDetails(path)[0]
+  }
+
+  private getErrorDetails(path: string): IErrorDetail[] {
+    let detail: IErrorDetail|null = null;
+    let nested: IErrorDetail|null = null;
     const details: IErrorDetail[] = [];
     for (let i = this._propNames.length - 1; i >= 0; i--) {
       const p = this._propNames[i];
       path += (typeof p === "number") ? `[${p}]` : (p ? `.${p}` : "");
       const message = this._messages[i];
       if (message) {
-        details.push({path, message});
+        nested = {path, message}
+        if (detail) {
+          detail.nested = [nested]
+        }
+        detail = nested
+        details.push(detail);
       }
     }
-    let detail: IErrorDetail|null = null;
-    for (let i = details.length - 1; i >= 0; i--) {
-      if (detail) { details[i].nested = [detail]; }
-      detail = details[i];
+    if (this._forks.length) {
+      if (detail) {
+        detail.nested = this._forks.map(fork => fork.getErrorDetail(path));
+      }
+      // Just keep one fork to preserve the old flat error message for now
+      details.push(...this._forks[0].getErrorDetails(path));
+    }
+
+    return details;
+  }
+
+  public fork(): IContext {
+    const ctx = new DetailContext();
+    this._forks.push(ctx);
+    return ctx;
+  }
+
+  public completeFork(): boolean {
```

Comment:
I had trouble following the logic and assumptions of `completeFork()` and `fork()`. It would help to document their purpose, and also the usage of `this._forks` member.

Reply:
Absolutely, there'll be lots of comments, I just put that off because the code was still experimental.